lib/write-set-tokens.ts

```diff
@@ -0,0 +1,98 @@
+import { types, Set, Range, Char, SetTokens } from './types';
+import * as sets from './sets-lookup';
+
+/**
+ * Takes character code and returns character to be displayed in a set
+ * @param {number} charCode Character code of set element
+ * @param {boolean} firstChar True if character is the first character of the set
+ * @param {boolean} lastChar True if character is the last character of the set
+ * @returns {string} The string for the sets character
+ */
+export function setChar(charCode: number, firstChar: boolean, lastChar = false): string {
+  // We only need to negate first internal '^' if token.not is false
+  return firstChar && charCode === 94 ? '\\^' :
+    charCode === 92 ? '\\\\' :
+      charCode === 93 ? '\\]' :
+        charCode === 45 && !lastChar ? '\\-' :
+          String.fromCharCode(charCode);
+}
+
+/**
+ * Test if a character set matches a 'set-lookup'
+ * @param {SetTokens} set The set to be tested
+ * @param {Record<string | number, boolean>} map The predefined 'set-lookup'
+ * @returns {boolean} True if the character set corresponds to the 'set-lookup'
+ */
+function sameSet(set: SetTokens, map: Record<string | number, boolean>): boolean {
+  for (const elem of set) {
+    if (elem.type === types.SET) {
+      return false;
+    }
+    const key = elem.type === types.CHAR ? elem.value : `${elem.from}-${elem.to}`;
+    if (map[key]) {
+      map[key] = false;
```

Comment:
to avoid checking for `len` of set outside of this function, delete the key here, then check that length of `map` is 0

Reply:
Are you sure? Currently, there *at most* one call to `isSameSet` per nested set, your suggestion means that, for instance, a whitespace nested set would first have to fail the lookups for `INTS`, `WORDS` and `NOTANYCHAR` so it would cause a performance decrease (albeit negligible in most cases)

Reply:
that's a good point. i just wish the length was retrieved from somewhere. but both the sets here and during tokenizing are created inside a function

Reply:
One option would be to add a `size` key to each of the maps; i.e.
```ts
export const INTS = (): Record<string | number, boolean | number> => ({
  '48-57': true,
  size: 1
});
```
but that solution feels messy to me.

Alternatively the `Map` constructor could be used because it has `size`; but from my understanding the initialisation would look like: 
```ts
function INTS() {
  const intsMap = new Map();
  intsMap.set('45-57', true);
  return intsMap;
}
```
i.e. having to cal `.set` on each entry, or,
```ts
function INTS() {
  const intsMap = new Map([['45-57', true]]);
  return intsMap;
}
```
and I am not sure how the time complexity compares to the current approach

Reply:
yeah, it would require creating the map beforehand too. it's fine the way it is